File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def sameTree(n1, n2):
            if not n1 and not n2:
                return True
            if not n1 or not n2:
                return False
            if n1.val != n2.val:
                return False
            return sameTree(n1.left, n2.left) and sameTree(n1.right, n2.right)

        def singleDfs(node):
            if not node and not subRoot:
                return True
            if not node:
                return False
            if node.val == subRoot.val:
                if sameTree(node, subRoot):
                    return True
            return singleDfs(node.left) or singleDfs(node.right)

        return singleDfs(root)
            


        

            
        

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int count = 0;

    public void dfs(TreeNode node, int max_value){
        if(node == null)
            return;
        if(node.val >= max_value){
            count++;
            max_value = node.val;
        }
        dfs(node.left, max_value);
        dfs(node.right, max_value);
    }
    public int goodNodes(TreeNode root) {
        dfs(root, root.val);
        return count;
    }
}
